### PR TITLE
Port SetAcceptance to the TCK

### PR DESCRIPTION
**Pull request by petraselmer**
This must only be merged when https://github.com/neo4j/neo4j/pull/7366 is merged

### CODE REVIEW

File: tck/features/SetAcceptance.feature [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+    And the side effects should be:
+      | +properties | 3 |
+
+  Scenario: Fail at runtime when the expression is neither a node nor a relationship
+    Given any graph
+    And parameters are:
+      | node | 42 |
+    When executing query:
+      """
+      SET (CASE WHEN true THEN {node} END).name = 'neo4j'
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: CASE expressions are not openCypher scope.

File: tck/features/SetAcceptance.feature [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+Feature: SetAcceptance
+
+  Scenario: Setting a node property to null removes the existing property
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: I don't think we can say `any` when we have a `MATCH` in the tested query.

File: tck/features/SetAcceptance.feature [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+      """
+      MATCH (n:A)
+      SET n.property1 = null
+      RETURN n
+      """
+    Then the result should be:
+      | n                    |
+      | (:A {property2: 46}) |
+    And the side effects should be:
+      | +properties | 1 |
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: I thought the side effects would've been one negative as well?

File: tck/features/SetAcceptance.feature [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      """
+      MATCH ()-[r]->()
+      SET r.property1 = null
+      RETURN r
+      """
+    Then the result should be:
+      | r                      |
+      | [:REL {property2: 24}] |
+    And the side effects should be:
+      | +properties | 1 |
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: The side effect should be just `-properties | 1`, I think.

File: tck/features/SetAcceptance.feature [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                    |
+      | (:A {property2: 46}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Setting a relationship property to null removes the existing property
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                      |
+      | (:A {name: 'Michael'}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Set a property to an expression
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching; same for all below scenarios.

File: tck/features/SetAcceptance.feature [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+  Scenario: Set a property by selecting the node using an expression
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)
+      """
+    When executing query:
+      """
+      MATCH (n:A)
+      SET (CASE WHEN true THEN n END).name = 'neo4j'
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: No CASE expressions in oC.

File: tck/features/SetAcceptance.feature [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+  Scenario: Set a property by selecting the node using a simple expression
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)
+      """
+    When executing query:
+      """
+      MATCH (n:A)
+      SET (n).name = 'neo4j'
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Are the parentheses intentional here?

File: tck/features/SetAcceptance.feature [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+  Scenario: Set a property by selecting the relationship using an expression
+    Given any graph
+    And having executed:
+      """
+      CREATE ()-[:REL]->()
+      """
+    When executing query:
+      """
+      MATCH ()-[r:REL]->()
+      SET (CASE WHEN true THEN r END).name = 'neo4j'
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: No CASE expressions in oC.

File: tck/features/SetAcceptance.feature [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+      MATCH (n)
+      WHERE n.name = 'Michael'
+      SET n.name = null
+      RETURN n
+      """
+    Then the result should be:
+      | n              |
+      | (:A {age: 35}) |
+    And the side effects should be:
+      | +properties | 1 |
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Should be a negative effect.

File: tck/features/SetAcceptance.feature [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+  Scenario: Add a label to a node
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)
+      """
+    When executing query:
+      """
+      MATCH (n:A)
+      SET n:FOO
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: The label should be `Foo`.

File: tck/features/SetAcceptance.feature [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+    Given any graph
+    And having executed:
+      """
+      CREATE (:A)
+      """
+    When executing query:
+      """
+      MATCH (n:A)
+      SET n.x = [1, 2, 3]
+      RETURN extract(i IN n.x | i/2.0) AS x
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: The `/` operator should have padding space around it.

Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Also, no `extract()`. Use a list comprehension, like so:

`RETURN [i IN n.x | i / 2.0]`

File: tck/features/SetAcceptance.feature [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                              |
+      | (:A {name: 'Andres was here'}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the node using a simple expression
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                    |
+      | (:A {name: 'neo4j'}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the node using an expression
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                    |
+      | (:A {name: 'neo4j'}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the relationship using a simple expression
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+    And having executed:
+      """
+      CREATE ()-[:REL]->()
+      """
+    When executing query:
+      """
+      MATCH ()-[r:REL]->()
+      SET (r).name = 'neo4j'
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Are the parentheses intentional here?

File: tck/features/SetAcceptance.feature [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN r
+      """
+    Then the result should be:
+      | r                      |
+      | [:REL {name: 'neo4j'}] |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Set a property by selecting the relationship using an expression
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN r
+      """
+    Then the result should be:
+      | r                      |
+      | [:REL {name: 'neo4j'}] |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Setting a property to null removes the property
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n              |
+      | (:A {age: 35}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Add a label to a node
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n        |
+      | (:A:FOO) |
+    And the side effects should be:
+      | +labels | 1 |
+
+  Scenario: Adding a list property
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 21]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                         |
+      | (:X {foo: 'A', bar: 'C'}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Retain old values when using +=
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+      """
+    Then the result should be:
+      | a.foo           |
+      | [1, 2, 3, 4, 5] |
+    And the side effects should be:
+      | +nodes      | 1 |
+      | +properties | 2 |
+
+  Scenario: Overwrite values when using +=
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature [hunk 23]
```diff
@@ -0,0 +1,332 @@
...
+      """
+      MATCH (n:X {foo: 'A'})
+      SET n += {foo: null}
+      RETURN n
+      """
+    Then the result should be:
+      | n               |
+      | (:X {bar: 'B'}) |
+    And the side effects should be:
+      | +properties | 1 |
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Should be a negative effect.

File: tck/features/SetAcceptance.feature [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+      """
+      MATCH (n:X {foo: 'A'})
+      SET n = {foo: 'B', baz: 'C'}
+      RETURN n
+      """
+    Then the result should be:
+      | n                         |
+      | (:X {foo: 'B', baz: 'C'}) |
+    And the side effects should be:
+      | +properties | 3 |
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Two + and one -

File: tck/features/SetAcceptance.feature [hunk 25]
```diff
@@ -0,0 +1,332 @@
...
+  Scenario: Fail at runtime when the expression is neither a node nor a relationship
+    Given any graph
+    And parameters are:
+      | node | 42 |
+    When executing query:
+      """
+      SET (CASE WHEN true THEN {node} END).name = 'neo4j'
+      RETURN count(*)
+      """
+    Then a ArgumentError should be raised at runtime: RequiresNodeOrRelationship
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: The problem is that the property access operator (`.`) does not work on the boolean value, right? That sounds to me like `IllegalArgumentValue`. A map would also be a valid value, and having `RequiresXorYorZ` type error details quickly gets out of hand I believe.

File: tck/features/SetAcceptance.feature [hunk 26]
```diff
@@ -0,0 +1,332 @@
...
+      RETURN n
+      """
+    Then the result should be:
+      | n                         |
+      | (:X {foo: 'A', bar: 'B'}) |
+    And the side effects should be:
+      | +properties | 1 |
+
+  Scenario: Explicit null values in a map remove old values
+    Given any graph
```
Comment (Mats-SX) [inline tck/features/SetAcceptance.feature]: Can't have `any graph` if we're matching.

File: tck/features/SetAcceptance.feature (hunk 8 above)
Comment (petraselmer) [inline tck/features/SetAcceptance.feature]: @Mats-SX yes, the original test has this query: https://github.com/neo4j/neo4j/blob/3.1/community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/SetAcceptanceTest.scala#L80

File: tck/features/SetAcceptance.feature (hunk 16 above)
Comment (petraselmer) [inline tck/features/SetAcceptance.feature]: @Mats-SX yes, the original query is here: https://github.com/neo4j/neo4j/blob/3.1/community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/SetAcceptanceTest.scala#L104

**Comment by petraselmer** [comment]
@Mats-SX thanks for reviewing. I have removed all tests containing `CASE`; these ought to be simulated and added later on in one iteration. Also, have hopefully addressed all of your other comments.

**Pull request merged**